File: binary-tree-maximum-path-sum.py
from __future__ import annotations
from typing import Optional

class TreeNode:
    def __init__(self, val: int = 0, left: Optional[TreeNode] = None, right: Optional[TreeNode] = None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        path_sum = -float("inf")

        def sum_without_split(node: Optional[TreeNode]) -> int:
            nonlocal path_sum
            if node is None: return 0

            left = max(sum_without_split(node.left), 0)
            right = max(sum_without_split(node.right), 0)

            path_sum = max(path_sum, left + node.val + right)
            return node.val + max(left, right)

        sum_without_split(root)
        return path_sum

File: binary-tree-maximum-path-sum.cpp
#include <utility>
#include <limits>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int maxPathSum(TreeNode* root) {
		pair<int, int> path_sum = find_path_sum(root);
		return max(path_sum.first, path_sum.second);
    }

	pair<int, int> find_path_sum(TreeNode* node) {
		if (node == nullptr) return pair<int, int>(0, INT_MIN);

		pair<int, int> left = find_path_sum(node->left);
		pair<int, int> right = find_path_sum(node->right);

		left.first = max(left.first, 0);
		right.first = max(right.first, 0);

		int sum_with_split = max(
			node->val + left.second + right.second,
			max(node->val, max(left.second, right.second))
		);

		int sum_without_split = node->val + max(left.first, right.first);

		return pair<int, int>(sum_without_split, sum_with_split);
	}
};
